File: web/src/lib/adapters/jismesh.ts
/**
 * 地域標準メッシュ（JIS X 0410）アダプター。
 *
 * レベル体系:
 *   "1"       1次メッシュ（約80km）  緯度 2/3°   × 経度 1°
 *   "2"       2次メッシュ（約10km）  緯度 1/12°  × 経度 1/8°
 *   "3"       3次メッシュ（約1km）   緯度 1/120° × 経度 1/80°
 *   "half"    2分の1メッシュ（約500m）
 *   "quarter" 4分の1メッシュ（約250m）
 *   "eighth"  8分の1メッシュ（約125m）
 *
 * QGIS プラグイン側 adapters/jismesh.py と同一ロジック（指示書 §16）。
 */

import type { GridAdapter, GridCell, Position } from "../types";
import { distanceM, polygonAreaM2 } from "../geo";

export type JismeshLevel = "1" | "2" | "3" | "half" | "quarter" | "eighth";

interface LevelSpec {
  latSpan: number;
  lonSpan: number;
  codeLength: number;
}

export const JISMESH_LEVELS: Record<JismeshLevel, LevelSpec> = {
  "1": { latSpan: 2 / 3, lonSpan: 1, codeLength: 4 },
  "2": { latSpan: 1 / 12, lonSpan: 1 / 8, codeLength: 6 },
  "3": { latSpan: 1 / 120, lonSpan: 1 / 80, codeLength: 8 },
  half: { latSpan: 1 / 240, lonSpan: 1 / 160, codeLength: 9 },
  quarter: { latSpan: 1 / 480, lonSpan: 1 / 320, codeLength: 10 },
  eighth: { latSpan: 1 / 960, lonSpan: 1 / 640, codeLength: 11 },
};

const LEVEL_ORDER: JismeshLevel[] = ["1", "2", "3", "half", "quarter", "eighth"];

/** 浮動小数点の桁落ちでセル境界上の点が隣セルに落ちるのを防ぐ許容量 */
const EPS = 1e-9;

function assertInRange(lon: number, lat: number): void {
  if (lat < 0 || lat >= 66.66 || lon < 100 || lon >= 180) {
    throw new Error(
      `地域標準メッシュの適用範囲外です (lon=${lon}, lat=${lat})`
    );
  }
}

export function levelOfCode(code: string): JismeshLevel {
  const found = LEVEL_ORDER.find(
    (lv) => JISMESH_LEVELS[lv].codeLength === code.length
  );
  if (!found || !/^\d+$/.test(code)) {
    throw new Error(`無効なメッシュコードです: ${code}`);
  }
  return found;
}

/** 経緯度からメッシュコードを計算する */
export function toMeshCode(
  lon: number,
  lat: number,
  level: JismeshLevel
): string {
  assertInRange(lon, lat);
  const p = Math.floor(lat * 1.5 + EPS);
  const u = Math.floor(lon + EPS) - 100;
  let code = `${String(p).padStart(2, "0")}${String(u).padStart(2, "0")}`;
  if (level === "1") return code;

  const q = Math.floor(lat * 12 + EPS) % 8;
  const v = Math.floor(lon * 8 + EPS) % 8;
  code += `${q}${v}`;
  if (level === "2") return code;

  const r = Math.floor(lat * 120 + EPS) % 10;
  const w = Math.floor(lon * 80 + EPS) % 10;
  code += `${r}${w}`;
  if (level === "3") return code;

  const depth = { half: 1, quarter: 2, eighth: 3 }[level];
  let latMul = 240;
  let lonMul = 160;
  for (let i = 0; i < depth; i++) {
    const latBit = Math.floor(lat * latMul + EPS) % 2;
    const lonBit = Math.floor(lon * lonMul + EPS) % 2;
    code += String(latBit * 2 + lonBit + 1);
    latMul *= 2;
    lonMul *= 2;
  }
  return code;
}

/** メッシュコードから南西端 [lonMin, latMin] を計算する */
export function codeToSouthWest(code: string): [number, number] {
  const level = levelOfCode(code);
  let latMin = parseInt(code.slice(0, 2), 10) / 1.5;
  let lonMin = parseInt(code.slice(2, 4), 10) + 100;
  if (level === "1") return [lonMin, latMin];

  latMin += parseInt(code[4], 10) / 12;
  lonMin += parseInt(code[5], 10) / 8;
  if (level === "2") return [lonMin, latMin];

  latMin += parseInt(code[6], 10) / 120;
  lonMin += parseInt(code[7], 10) / 80;
  if (level === "3") return [lonMin, latMin];

  let latSpan = 1 / 240;
  let lonSpan = 1 / 160;
  for (let i = 8; i < code.length; i++) {
    const d = parseInt(code[i], 10) - 1;
    if (d < 0 || d > 3) throw new Error(`無効なメッシュコードです: ${code}`);
    latMin += (d >> 1) * latSpan;
    lonMin += (d & 1) * lonSpan;
    latSpan /= 2;
    lonSpan /= 2;
  }
  return [lonMin, latMin];
}

function buildCell(code: string): GridCell {
  const level = levelOfCode(code);
  const { latSpan, lonSpan } = JISMESH_LEVELS[level];
  const [lonMin, latMin] = codeToSouthWest(code);
  const lonMax = lonMin + lonSpan;
  const latMax = latMin + latSpan;
  const ring: Position[] = [
    [lonMin, latMin],
    [lonMax, latMin],
    [lonMax, latMax],
    [lonMin, latMax],
    [lonMin, latMin],
  ];
  const center: [number, number] = [
    lonMin + lonSpan / 2,
    latMin + latSpan / 2,
  ];
  const widthM = distanceM([lonMin, center[1]], [lonMax, center[1]]);
  const heightM = distanceM([center[0], latMin], [center[0], latMax]);
  const idx = LEVEL_ORDER.indexOf(level);
  return {
    system: "jismesh",
    id: code,
    level,
    geometry: { type: "Polygon", coordinates: [ring] },
    center,
    areaM2: polygonAreaM2([ring]),
    widthM,
    heightM,
    parentId: idx > 0 ? parentOf(code) : undefined,
    childCount: level === "1" ? 64 : level === "2" ? 100 : level === "eighth" ? 0 : 4,
  };
}

function parentOf(code: string): string {
  const level = levelOfCode(code);
  const idx = LEVEL_ORDER.indexOf(level);
  if (idx === 0) throw new Error("1次メッシュに親はありません");
  const parentLevel = LEVEL_ORDER[idx - 1];
  return code.slice(0, JISMESH_LEVELS[parentLevel].codeLength);
}

function childrenOf(code: string): string[] {
  const level = levelOfCode(code);
  const idx = LEVEL_ORDER.indexOf(level);
  if (idx === LEVEL_ORDER.length - 1) return [];
  const childLevel = LEVEL_ORDER[idx + 1];
  const out: string[] = [];
  if (level === "1") {
    for (let q = 0; q < 8; q++)
      for (let v = 0; v < 8; v++) out.push(`${code}${q}${v}`);
  } else if (level === "2") {
    for (let r = 0; r < 10; r++)
      for (let w = 0; w < 10; w++) out.push(`${code}${r}${w}`);
  } else {
    for (let d = 1; d <= 4; d++) out.push(`${code}${d}`);
  }
  void childLevel;
  return out;
}

function neighborsOf(code: string): string[] {
  const level = levelOfCode(code);
  const { latSpan, lonSpan } = JISMESH_LEVELS[level];
  const [lonMin, latMin] = codeToSouthWest(code);
  const cLon = lonMin + lonSpan / 2;
  const cLat = latMin + latSpan / 2;
  const out: string[] = [];
  for (let dy = -1; dy <= 1; dy++) {
    for (let dx = -1; dx <= 1; dx++) {
      if (dx === 0 && dy === 0) continue;
      try {
        out.push(toMeshCode(cLon + dx * lonSpan, cLat + dy * latSpan, level));
      } catch {
        // 適用範囲外は無視
      }
    }
  }
  return out;
}

export const jismeshAdapter: GridAdapter = {
  system: "jismesh",
  displayName: "地域標準メッシュ",
  levels: [
    { value: "1", label: "1次メッシュ（約80km）" },
    { value: "2", label: "2次メッシュ（約10km）" },
    { value: "3", label: "3次メッシュ（約1km）" },
    { value: "half", label: "2分の1メッシュ（約500m）" },
    { value: "quarter", label: "4分の1メッシュ（約250m）" },
    { value: "eighth", label: "8分の1メッシュ（約125m）" },
  ],
  defaultLevel: "3",

  pointToCell(lon, lat, level) {
    return buildCell(toMeshCode(lon, lat, level as JismeshLevel));
  },

  cellToGeometry(id) {
    return buildCell(id);
  },

  getParent(id) {
    return levelOfCode(id) === "1" ? null : parentOf(id);
  },

  getChildren(id) {
    return childrenOf(id);
  },

  getNeighbors(id) {
    return neighborsOf(id);
  },

  cellsForBounds(bounds, level) {
    const [minLon, minLat, maxLon, maxLat] = bounds;
    const { latSpan, lonSpan } = JISMESH_LEVELS[level as JismeshLevel];
    const lat0 = Math.max(minLat, 0);
    const lat1 = Math.min(maxLat, 66.66 - latSpan / 2);
    const lon0 = Math.max(minLon, 100);
    const lon1 = Math.min(maxLon, 180 - lonSpan / 2);
    if (lat0 > lat1 || lon0 > lon1) return [];
    const cells: GridCell[] = [];
    const iMin = Math.floor(lat0 / latSpan + EPS);
    const iMax = Math.floor(lat1 / latSpan + EPS);
    const jMin = Math.floor(lon0 / lonSpan + EPS);
    const jMax = Math.floor(lon1 / lonSpan + EPS);
    if ((iMax - iMin + 1) * (jMax - jMin + 1) > 100000) {
      throw new Error("セル数が多すぎます。レベルを下げてください。");
    }
    for (let i = iMin; i <= iMax; i++) {
      for (let j = jMin; j <= jMax; j++) {
        const lat = (i + 0.5) * latSpan;
        const lon = (j + 0.5) * lonSpan;
        cells.push(
          buildCell(toMeshCode(lon, lat, level as JismeshLevel))
        );
      }
    }
    return cells;
  },
};

File: qgis-plugin/grid_interoperability/adapters/jismesh.py
# -*- coding: utf-8 -*-
"""地域標準メッシュ（JIS X 0410）アダプター。

Web版 src/lib/adapters/jismesh.ts と同一ロジック。
座標の丸め規則: セル境界上の点は EPS 付き floor により
北東側（値の大きい側）のセルに属する（半開区間 [下端, 上端)）。
"""

import math
from typing import List, Optional, Tuple

from ..core.geometry import distance_m, polygon_area_m2, rect_ring
from ..core.models import GridCell

SYSTEM = "jismesh"

# level -> (latSpan, lonSpan, codeLength)
LEVELS = {
    "1": (2.0 / 3.0, 1.0, 4),
    "2": (1.0 / 12.0, 1.0 / 8.0, 6),
    "3": (1.0 / 120.0, 1.0 / 80.0, 8),
    "half": (1.0 / 240.0, 1.0 / 160.0, 9),
    "quarter": (1.0 / 480.0, 1.0 / 320.0, 10),
    "eighth": (1.0 / 960.0, 1.0 / 640.0, 11),
}

LEVEL_ORDER = ["1", "2", "3", "half", "quarter", "eighth"]

LEVEL_LABELS = {
    "1": "1次メッシュ（約80km）",
    "2": "2次メッシュ（約10km）",
    "3": "3次メッシュ（約1km）",
    "half": "2分の1メッシュ（約500m）",
    "quarter": "4分の1メッシュ（約250m）",
    "eighth": "8分の1メッシュ（約125m）",
}

DEFAULT_LEVEL = "3"

# 浮動小数点誤差でセル境界上の点が隣セルへ落ちるのを防ぐ許容量（Web版と同値）
EPS = 1e-9


def _assert_in_range(lon: float, lat: float) -> None:
    if lat < 0 or lat >= 66.66 or lon < 100 or lon >= 180:
        raise ValueError(
            "地域標準メッシュの適用範囲外です (lon=%s, lat=%s)" % (lon, lat)
        )


def level_of_code(code: str) -> str:
    if not code.isdigit():
        raise ValueError("無効なメッシュコードです: %s" % code)
    for level in LEVEL_ORDER:
        if LEVELS[level][2] == len(code):
            return level
    raise ValueError("無効なメッシュコードです: %s" % code)


def to_mesh_code(lon: float, lat: float, level: str) -> str:
    _assert_in_range(lon, lat)
    p = math.floor(lat * 1.5 + EPS)
    u = math.floor(lon + EPS) - 100
    code = "%02d%02d" % (p, u)
    if level == "1":
        return code

    q = math.floor(lat * 12 + EPS) % 8
    v = math.floor(lon * 8 + EPS) % 8
    code += "%d%d" % (q, v)
    if level == "2":
        return code

    r = math.floor(lat * 120 + EPS) % 10
    w = math.floor(lon * 80 + EPS) % 10
    code += "%d%d" % (r, w)
    if level == "3":
        return code

    depth = {"half": 1, "quarter": 2, "eighth": 3}[level]
    lat_mul, lon_mul = 240, 160
    for _ in range(depth):
        lat_bit = math.floor(lat * lat_mul + EPS) % 2
        lon_bit = math.floor(lon * lon_mul + EPS) % 2
        code += str(lat_bit * 2 + lon_bit + 1)
        lat_mul *= 2
        lon_mul *= 2
    return code


def code_to_south_west(code: str) -> Tuple[float, float]:
    level = level_of_code(code)
    lat_min = int(code[0:2]) / 1.5
    lon_min = int(code[2:4]) + 100.0
    if level == "1":
        return lon_min, lat_min

    lat_min += int(code[4]) / 12.0
    lon_min += int(code[5]) / 8.0
    if level == "2":
        return lon_min, lat_min

    lat_min += int(code[6]) / 120.0
    lon_min += int(code[7]) / 80.0
    if level == "3":
        return lon_min, lat_min

    lat_span, lon_span = 1.0 / 240.0, 1.0 / 160.0
    for ch in code[8:]:
        d = int(ch) - 1
        if d < 0 or d > 3:
            raise ValueError("無効なメッシュコードです: %s" % code)
        lat_min += (d >> 1) * lat_span
        lon_min += (d & 1) * lon_span
        lat_span /= 2
        lon_span /= 2
    return lon_min, lat_min


def cell_from_code(code: str) -> GridCell:
    level = level_of_code(code)
    lat_span, lon_span, _ = LEVELS[level]
    lon_min, lat_min = code_to_south_west(code)
    lon_max = lon_min + lon_span
    lat_max = lat_min + lat_span
    ring = rect_ring(lon_min, lat_min, lon_max, lat_max)
    center = (lon_min + lon_span / 2, lat_min + lat_span / 2)
    idx = LEVEL_ORDER.index(level)
    return GridCell(
        system=SYSTEM,
        id=code,
        level=level,
        rings=[ring],
        center=center,
        area_m2=polygon_area_m2([ring]),
        width_m=distance_m((lon_min, center[1]), (lon_max, center[1])),
        height_m=distance_m((center[0], lat_min), (center[0], lat_max)),
        parent_id=get_parent(code) if idx > 0 else None,
        child_count=(
            64 if level == "1" else 100 if level == "2" else 0 if level == "eighth" else 4
        ),
    )


def point_to_cell(lon: float, lat: float, level, height_m: float = 0.0) -> GridCell:
    return cell_from_code(to_mesh_code(lon, lat, str(level)))


def cell_to_geometry(code: str) -> GridCell:
    return cell_from_code(code)


def get_parent(code: str) -> Optional[str]:
    level = level_of_code(code)
    idx = LEVEL_ORDER.index(level)
    if idx == 0:
        return None
    parent_level = LEVEL_ORDER[idx - 1]
    return code[: LEVELS[parent_level][2]]


def get_children(code: str) -> List[str]:
    level = level_of_code(code)
    idx = LEVEL_ORDER.index(level)
    if idx == len(LEVEL_ORDER) - 1:
        return []
    if level == "1":
        return ["%s%d%d" % (code, q, v) for q in range(8) for v in range(8)]
    if level == "2":
        return ["%s%d%d" % (code, r, w) for r in range(10) for w in range(10)]
    return ["%s%d" % (code, d) for d in range(1, 5)]


def get_neighbors(code: str) -> List[str]:
    level = level_of_code(code)
    lat_span, lon_span, _ = LEVELS[level]
    lon_min, lat_min = code_to_south_west(code)
    c_lon = lon_min + lon_span / 2
    c_lat = lat_min + lat_span / 2
    out = []
    for dy in (-1, 0, 1):
        for dx in (-1, 0, 1):
            if dx == 0 and dy == 0:
                continue
            try:
                out.append(
                    to_mesh_code(c_lon + dx * lon_span, c_lat + dy * lat_span, level)
                )
            except ValueError:
                pass
    return out


def cells_for_bounds(bounds, level) -> List[GridCell]:
    level = str(level)
    min_lon, min_lat, max_lon, max_lat = bounds
    lat_span, lon_span, _ = LEVELS[level]
    lat0 = max(min_lat, 0.0)
    lat1 = min(max_lat, 66.66 - lat_span / 2)
    lon0 = max(min_lon, 100.0)
    lon1 = min(max_lon, 180.0 - lon_span / 2)
    if lat0 > lat1 or lon0 > lon1:
        return []
    i_min = math.floor(lat0 / lat_span + EPS)
    i_max = math.floor(lat1 / lat_span + EPS)
    j_min = math.floor(lon0 / lon_span + EPS)
    j_max = math.floor(lon1 / lon_span + EPS)
    if (i_max - i_min + 1) * (j_max - j_min + 1) > 1000000:
        raise ValueError("セル数が多すぎます。レベルを下げてください。")
    cells = []
    for i in range(i_min, i_max + 1):
        for j in range(j_min, j_max + 1):
            lat = (i + 0.5) * lat_span
            lon = (j + 0.5) * lon_span
            cells.append(cell_from_code(to_mesh_code(lon, lat, level)))
    return cells


def levels_for_ui():
    return [(lv, LEVEL_LABELS[lv]) for lv in LEVEL_ORDER]
